wallet/src/main/scala/org/bitcoins/wallet/models/AddressTagDAO.scala

```diff
@@ -0,0 +1,89 @@
+package org.bitcoins.wallet.models
+
+import org.bitcoins.core.config.NetworkParameters
+import org.bitcoins.core.crypto.{DoubleSha256Digest, DoubleSha256DigestBE}
+import org.bitcoins.core.protocol.BitcoinAddress
+import org.bitcoins.core.wallet.utxo.{AddressTag, AddressTagType}
+import org.bitcoins.db.{CRUD, SlickUtil}
+import org.bitcoins.wallet.config.WalletAppConfig
+import slick.jdbc.SQLiteProfile.api._
+import slick.lifted.TableQuery
+
+import scala.concurrent.{ExecutionContext, Future}
+
+case class AddressTagDAO()(
+    implicit ec: ExecutionContext,
+    config: WalletAppConfig
+) extends CRUD[AddressTagDb, BitcoinAddress] {
+  import org.bitcoins.db.DbCommonsColumnMappers._
+
+  override val table: TableQuery[AddressTagTable] = TableQuery[AddressTagTable]
+  private val spendingInfoTable: TableQuery[SpendingInfoTable] =
+    TableQuery[SpendingInfoTable]
+
+  override def createAll(
+      ts: Vector[AddressTagDb]): Future[Vector[AddressTagDb]] =
+    SlickUtil.createAllNoAutoInc(ts, database, table)
+
+  /** Finds the rows that correlate to the given primary keys */
+  override def findByPrimaryKeys(
+      addresses: Vector[BitcoinAddress]): Query[Table[_], AddressTagDb, Seq] =
+    table.filter(_.address.inSet(addresses))
+
+  override def findByPrimaryKey(
+      address: BitcoinAddress): Query[Table[_], AddressTagDb, Seq] = {
+    table.filter(_.address === address)
+  }
+
+  override def findAll(
+      ts: Vector[AddressTagDb]): Query[Table[_], AddressTagDb, Seq] =
+    findByPrimaryKeys(ts.map(_.address))
+
+  def findByAddress(address: BitcoinAddress): Future[Vector[AddressTagDb]] = {
+    val query = table.filter(_.address === address)
+
+    database.run(query.result).map(_.toVector)
+  }
+
+  def findByTag(tag: AddressTag): Future[Vector[AddressTagDb]] = {
+    val byTag = table.filter(_.tag === tag)
+
+    val query = tag.tagTypeOpt match {
+      case Some(tagType) =>
+        byTag.filter(_.tagType === tagType)
+      case None =>
+        byTag
+    }
+
+    database.run(query.result).map(_.toVector)
+  }
+
+  def findByTagType(tagType: AddressTagType): Future[Vector[AddressTagDb]] = {
+    val query = table.filter(_.tagType === tagType)
+
+    database.run(query.result).map(_.toVector)
+  }
+
+  def findTx(
+      txIdBE: DoubleSha256DigestBE,
+      network: NetworkParameters): Future[Vector[AddressTagDb]] = {
+    val infoQuery = spendingInfoTable.filter(_.txid === txIdBE)
+    val spendingInfosF = database.runVec(infoQuery.result)
+
+    spendingInfosF.flatMap { spendingInfos =>
+      if (spendingInfos.isEmpty) {
+        Future.successful(Vector.empty)
+      } else {
+        val firstSpk = spendingInfos.head.output.scriptPubKey
```

Comment:
Isn't this assuming we don't have multiple utxos on the same txid? Probably not safe